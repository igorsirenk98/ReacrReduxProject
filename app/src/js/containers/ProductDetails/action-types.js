export const FETCH_PRODUCT_DETAILS_START = 'FETCH_PRODUCT_DETAILS_START';
export const FETCH_PRODUCT_DETAILS_RESULT = 'FETCH_PRODUCT_DETAILS_RESULT';