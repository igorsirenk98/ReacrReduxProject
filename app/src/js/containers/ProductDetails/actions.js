import {
    FETCH_PRODUCT_DETAILS_START,
    FETCH_PRODUCT_DETAILS_RESULT
} from './action-types';

import callApi from '../../utils/callApi';

export const fetchProductDetailsStart = () => ({
    type: FETCH_PRODUCT_DETAILS_START,
});

export const fetchProductDetailsResult = (payload, error) => ({
    type: FETCH_PRODUCT_DETAILS_RESULT,
    payload,
    error
});

export function fetchProductDetails(id) {
    const url = `products/productId=${id}`;
    
    return (dispatch) => {
		return dispatch(callApi([fetchProductDetailsStart, fetchProductDetailsResult], url));
	}
}
