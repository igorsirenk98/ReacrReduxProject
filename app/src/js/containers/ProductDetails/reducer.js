import {
	FETCH_PRODUCT_DETAILS_START,
	FETCH_PRODUCT_DETAILS_RESULT
} from './action-types';

const initialState = {
	payload: {
		product: {}
	},
	error: false
};

const productDetails = (state = initialState, action) => {
	switch (action.type) {
		case FETCH_PRODUCT_DETAILS_START: {
			return {
				...state,
				error: false
			};
		}
		case FETCH_PRODUCT_DETAILS_RESULT: {
			return {
				...state,
				payload: action.payload,
				error: action.error
			};
		}
		default:
			return state;
	}
};

export default productDetails;
