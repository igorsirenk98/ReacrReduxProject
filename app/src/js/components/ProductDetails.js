import React, { Component } from 'react';
import { connect } from 'react-redux';

import { ProductCard } from './basicComponents/ProductCard';

import { fetchProductById } from '../actions/ProductById';

const mapStateToProps = state => ({
    product: state.productById.product,
    error: state.productById.error
});

class ProductDetails extends Component {
    componentDidMount() {
        const { productId } = this.props.match.params;
        const { error } = this.props;

        this.props.dispatch(fetchProductById(productId));
    }

    render() {
        const { product, error } = this.props;

        if (error) {
            return <div>Error! {error.message}</div>;
        }

        return (
            <ProductCard product={product} />
        )
    }
}

export default connect(mapStateToProps)(ProductDetails);