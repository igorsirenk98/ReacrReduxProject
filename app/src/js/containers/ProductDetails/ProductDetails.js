import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import CssBaseline from '@material-ui/core/CssBaseline';
import Container from '@material-ui/core/Container';

import ProductCardDetailed from '../../components/ProductDetailsCard/ProductDetailsCard';
import Loading from '../controls/Loading/Loading';
import ErrorMessage from '../../components/controls/ErrorMessage';

import { fetchProductDetails } from './actions';

const mapStateToProps = state => ({
    payload: state.productDetails.payload,
    isLoading: state.loadingAnimation.payload.isLoading,
    error: state.productDetails.error,
});

const mapDispatchToProps = dispatch => {
    return bindActionCreators({
        fetchProductDetails
    }, dispatch);
};

export class ProductDetails extends Component {
    componentDidMount() {
        const { productId } = this.props.match.params;

        this.props.fetchProductDetails(productId);
    }

    render() {
        const { payload, isLoading, error } = this.props;
        const { product } = payload;

        if (error) {
            return <ErrorMessage error={payload} />;
        } else if (isLoading) {
            return (
                <>
                    <CssBaseline />
                    <Container className="container" fixed>
                        <div className="loading-container">
                            <Loading />
                        </div>
                    </Container>
                </>
            );
        }

        return (
            <>
                <CssBaseline />
                <Container className="container" fixed>
                    <ProductCardDetailed product={product} />
                </Container>
            </>
        );
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(ProductDetails);
