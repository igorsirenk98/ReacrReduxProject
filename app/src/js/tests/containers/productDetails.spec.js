import React from 'react';
import { shallow } from 'enzyme';
import { ProductDetails } from '../../containers/ProductDetails/ProductDetails';

describe('ProductDetails container', () => {
    const props = {
        payload: {
            product: {}
        },
        match: {
            params: 1
        },
        isLoading: false,
        error: false,
        fetchProductDetails: () => { },
    }

    describe('ProductDetails initial', () => {
        const productDetails = shallow(<ProductDetails {...props} />);

        it('renders correctly', () => {
            expect(productDetails.debug()).toMatchSnapshot();
        })
    })

    describe('ProductDetails isLoading', () => {
        const nextProps = {
            ...props,
            payload: {
                product: {
                    ...props.product,
                }
            },
            isLoading: true,
        }

        const productDetails = shallow(<ProductDetails {...nextProps} />);

        it('renders correctly', () => {
            expect(productDetails.debug()).toMatchSnapshot();
        })
    })

    describe('ProductDetails error', () => {
        const nextProps = {
            ...props,
            payload: {
                error: {
                    message: 'error',
                }
            },
            error: true,
        }

        const productDetails = shallow(<ProductDetails {...nextProps} />);

        it('renders correctly', () => {
            expect(productDetails.debug()).toMatchSnapshot();
        })
    })

    describe('ProductDetails with data', () => {
        const nextProps = {
            ...props,
            payload: {
                product: {}
            }
        }

        const productDetails = shallow(<ProductDetails {...nextProps} />);

        it('renders correctly', () => {
            expect(productDetails.debug()).toMatchSnapshot();
        })
    })
})

