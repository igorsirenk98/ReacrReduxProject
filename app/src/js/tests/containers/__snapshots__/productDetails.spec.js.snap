// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDetails container ProductDetails error renders correctly 1`] = `"<ErrorMessage error={{...}} />"`;

exports[`ProductDetails container ProductDetails initial renders correctly 1`] = `
"<Fragment>
  <CssBaseline />
  <WithStyles(ForwardRef(Container)) className=\\"container\\" fixed={true}>
    <ProductDetailsCard product={{...}} />
  </WithStyles(ForwardRef(Container))>
</Fragment>"
`;

exports[`ProductDetails container ProductDetails isLoading renders correctly 1`] = `
"<Fragment>
  <CssBaseline />
  <WithStyles(ForwardRef(Container)) className=\\"container\\" fixed={true}>
    <div className=\\"loading-container\\">
      <Connect(Loading) />
    </div>
  </WithStyles(ForwardRef(Container))>
</Fragment>"
`;

exports[`ProductDetails container ProductDetails with data renders correctly 1`] = `
"<Fragment>
  <CssBaseline />
  <WithStyles(ForwardRef(Container)) className=\\"container\\" fixed={true}>
    <ProductDetailsCard product={{...}} />
  </WithStyles(ForwardRef(Container))>
</Fragment>"
`;
